main/src/ca/uwaterloo/flix/language/phase/ConstraintResolution.scala

```diff
@@ -0,0 +1,716 @@
+          case Type.Var(sym, _) if renv0.isFlexible(sym) =>
+            Result.Ok((List(TypingConstraint.Class(clazz, t, loc)), progress))
+          // Case 2: Assoc type. It might be resolved later.
+          case _: Type.AssocType =>
+            Result.Ok((List(TypingConstraint.Class(clazz, t, loc)), progress))
+          // Case 3: Something rigid (const or rigid var). We can look this up immediately.
+          case _ =>
+            // we mark t's tvars as rigid so we get the substitution in the right direction
+            val renv = t.typeVars.map(_.sym).foldLeft(renv0)(_.markRigid(_))
+            val insts = classEnv(clazz).instances
+            // find the first (and only) instance that matches
+            val tconstrsOpt = ListOps.findMap(insts) {
+              inst =>
+                Unification.unifyTypes(t, inst.tpe, renv).toOption.flatMap {
+                  case (subst, Nil) => Some(inst.tconstrs.map(subst.apply))
+                  case (_, _ :: _) => None // if we have leftover constraints then it didn't actually unify
+                }
+            }
+            tconstrsOpt match {
+              case None =>
+                t.baseType match {
+                  // If it's a var, it's ok. It may be substituted later to a type we can reduce.
+                  // Or it might be part of the signature an expected constraint.
+                  case Type.Var(sym, loc) => Result.Ok(List(TypingConstraint.Class(clazz, t, loc)), progress)
+                  // If it's an associated type, it's ok. It may be reduced later to a concrete type.
+                  case _: Type.AssocType => Result.Ok(List(TypingConstraint.Class(clazz, t, loc)), progress)
+                  // Otherwise it's a problem.
+                  case _ => Result.Err(TypeError.MissingInstance(clazz, tpe0, renv, loc))
+                }
+              case Some(tconstrs) =>
+                // simplify all the implied constraints
+                Result.traverse(tconstrs) {
+                  case Ast.TypeConstraint(Ast.TypeConstraint.Head(c, _), arg, _) =>
+                    resolveClassConstraint(c, arg, classEnv, eqEnv, renv0, loc)
+                } map {
+                  case res =>
+                    val cs = res.flatMap { case (c, _) => c }
+                    (cs, true)
+                }
+            }
+        }
+    }
+  }
+
+  /**
+    * Attempts to resolve the given type constraints under the given environments,
+    * and with the given initial substitution.
+    *
+    * The initial substitution should come from e.g., formal parameter type ascriptions.
+    *
+    * Returns a reduction result containing a substitution and any leftover constraints.
+    */
+  private def resolve(constrs: List[TypingConstraint], subst0: Substitution, renv: RigidityEnv, cenv: Map[Symbol.ClassSym, Ast.ClassContext], eqEnv: ListMap[Symbol.AssocTypeSym, Ast.AssocTypeDef])(implicit flix: Flix): Result[ResolutionResult, TypeError] = {
```

Comment:
I was wondering: the cev and eqEnv stay fixed right? So maybe we should make them implicit through all functions?